File: src/parser.ts
import type { GitHubRepoItem, GitHubTreeItem, ThemeEntry, ThemeVariant } from "./types.js";

const COLORS_FILE_PATTERN = /^colors\/([^/]+)\.(vim|lua)$/;

const SUFFIXES_TO_STRIP = [
  ".nvim",
  ".vim",
  ".lua",
  "-nvim",
  "_nvim",
  "-vim",
  "_vim",
  "-colorscheme",
];

function sanitizeRepoName(repoName: string): string {
  let candidate = repoName.toLowerCase().trim();
  for (const suffix of SUFFIXES_TO_STRIP) {
    if (candidate.endsWith(suffix) && candidate.length > suffix.length) {
      candidate = candidate.slice(0, -suffix.length);
    }
  }
  candidate = candidate.replace(/^[-_]+|[-_]+$/g, "");
  return candidate;
}

export function normalizeThemeName(fullRepo: string): string {
  const slashIndex = fullRepo.indexOf("/");
  const owner = slashIndex >= 0 ? fullRepo.slice(0, slashIndex) : "";
  const repoName = slashIndex >= 0 ? fullRepo.slice(slashIndex + 1) : fullRepo;

  const cleanedRepo = sanitizeRepoName(repoName);
  const invalidNames = new Set(["", "nvim", "vim", "neovim", "theme", "colorscheme"]);

  if (invalidNames.has(cleanedRepo)) {
    const fallback = sanitizeRepoName(owner);
    if (fallback) {
      return fallback;
    }
  }

  return cleanedRepo || sanitizeRepoName(owner) || "theme";
}

export function extractColorschemes(treeItems: GitHubTreeItem[]): string[] {
  const colors = new Set<string>();

  for (const item of treeItems) {
    if (item.type !== "blob") {
      continue;
    }

    const match = item.path.match(COLORS_FILE_PATTERN);
    if (!match || !match[1]) {
      continue;
    }

    const colorscheme = match[1].trim();
    if (colorscheme) {
      colors.add(colorscheme);
    }
  }

  return Array.from(colors).sort();
}

function pickBaseColorscheme(themeName: string, colors: string[]): string {
  if (colors.length === 0) {
    return themeName;
  }

  const preferred = new Set([
    themeName,
    themeName.replace(/-/g, "_"),
    themeName.replace(/_/g, "-"),
  ]);

  for (const candidate of colors) {
    if (preferred.has(candidate)) {
      return candidate;
    }
  }

  for (const candidate of colors) {
    if (!candidate.includes("-") && !candidate.includes("_")) {
      return candidate;
    }
  }

  return colors[0]!;
}

export function buildEntry(repoPayload: GitHubRepoItem, colorschemes: string[]): ThemeEntry {
  const fullName = repoPayload.full_name;
  if (!fullName || !fullName.includes("/")) {
    throw new Error("invalid repository payload");
  }

  const themeName = normalizeThemeName(fullName);
  const baseColorscheme = pickBaseColorscheme(themeName, colorschemes);

  const variants: ThemeVariant[] = colorschemes
    .filter((c) => c !== baseColorscheme)
    .map((value) => ({
      name: value,
      colorscheme: value,
    }));

  const topics = Array.isArray(repoPayload.topics)
    ? repoPayload.topics.filter((t): t is string => typeof t === "string" && t.length > 0)
    : [];

  const entry: ThemeEntry = {
    name: themeName,
    repo: fullName,
    colorscheme: baseColorscheme,
    description: repoPayload.description ?? "",
    stars: repoPayload.stargazers_count ?? 0,
    topics,
    updated_at: repoPayload.updated_at ?? "",
    archived: repoPayload.archived ?? false,
    disabled: repoPayload.disabled ?? false,
  };

  if (variants.length > 0) {
    entry.variants = variants;
  }

  return entry;
}

File: internal/theme/parser.go
package theme

import (
	"regexp"
	"sort"
	"strings"

	"github.com/raulcorreia7/theme-browser-registry/internal/gh"
)

// Ported verbatim from src/sync/parser.ts extractColorschemes/buildEntry.

var reColorsFile = regexp.MustCompile(`^colors/([^/]+)\.(vim|lua)$`)

// ExtractColorschemes collects colorscheme base names from a repo git tree:
// blobs matching ^colors/<name>.(vim|lua)$, deduped and sorted.
func ExtractColorschemes(items []gh.TreeItem) []string {
	colors := map[string]bool{}
	for _, item := range items {
		if item.Type != "blob" {
			continue
		}
		m := reColorsFile.FindStringSubmatch(item.Path)
		if m == nil || m[1] == "" {
			continue
		}
		name := strings.TrimSpace(m[1])
		if name != "" {
			colors[name] = true
		}
	}
	out := make([]string, 0, len(colors))
	for name := range colors {
		out = append(out, name)
	}
	sort.Strings(out)
	return out
}

func pickBaseColorscheme(themeName string, colors []string) string {
	if len(colors) == 0 {
		return themeName
	}

	preferred := map[string]bool{
		themeName:                               true,
		strings.ReplaceAll(themeName, "-", "_"): true,
		strings.ReplaceAll(themeName, "_", "-"): true,
	}

	for _, candidate := range colors {
		if preferred[candidate] {
			return candidate
		}
	}
	for _, candidate := range colors {
		if !strings.Contains(candidate, "-") && !strings.Contains(candidate, "_") {
			return candidate
		}
	}
	return colors[0]
}

// BuildEntry converts a repository payload plus its tree-derived colorschemes
// into an index Entry. Errors when the full_name is not "owner/repo".
func BuildEntry(repo gh.RepoItem, colorschemes []string) (Entry, error) {
	fullName := repo.FullName
	if !strings.Contains(fullName, "/") {
		return Entry{}, errInvalidRepoPayload
	}

	themeName := NormalizeThemeName(fullName)
	baseColorscheme := pickBaseColorscheme(themeName, colorschemes)

	var variants []Variant
	for _, value := range colorschemes {
		if value == baseColorscheme {
			continue
		}
		v := Variant{Name: value, Colorscheme: value}
		if mode := InferModeFromColorscheme(value); mode != nil {
			v.Mode = *mode
		}
		variants = append(variants, v)
	}

	topics := make([]string, 0, len(repo.Topics))
	for _, t := range repo.Topics {
		if t != "" {
			topics = append(topics, t)
		}
	}
	if len(topics) == 0 {
		topics = nil
	}

	description := ""
	if repo.Description != nil {
		description = *repo.Description
	}

	entry := Entry{
		Name:        themeName,
		Repo:        fullName,
		Colorscheme: baseColorscheme,
		Description: description,
		Stars:       repo.Stargazers,
		Topics:      topics,
		UpdatedAt:   repo.UpdatedAt,
		Archived:    repo.Archived,
		Disabled:    repo.Disabled,
	}
	if len(variants) > 0 {
		entry.Variants = variants
	}
	return entry, nil
}
